[FSDP] Move the sharded_state_dict logic to the post hook to avoid OOM

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82613

The original implementation put the call of `_summon_full_params()` in `state_dict()`. However, because `state_dict()` is recursive, `_summon_full_params()` will also behave like the recursive version even if recursive is set to False. This PR put the logic in the post hook to solve the OOM issue.

Differential Revision: [D38329396](https://our.internmc.facebook.com/intern/diff/D38329396/)

fegin committed:
[FSDP] Move the sharded_state_dict logic to the post hook to avoid OOM

The original implementation put the call of `_summon_full_params()` in `state_dict()`. However, because `state_dict()` is recursive, `_summon_full_params()` will also behave like the recursive version even if recursive is set to False. This PR put the logic in the post hook to solve the OOM issue.

Differential Revision: [D38329396](https://our.internmc.facebook.com/intern/diff/D38329396/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1983,19 +1983,25 @@ def _sharded_post_state_dict_hook(
+                state_dict[fqn] = init_from_local_shards(
+                    local_shards, param.size(), process_group=self.process_group
+                )  # type: ignore[assignment]
+        state_dict.pop(f"{prefix}{FLAT_PARAM}")
```
I'm assuming that this is removing the key checkpointed by the `super().state_dict()` call?

fegin replied:
Yes

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2103,24 +2109,19 @@ def state_dict(self, *args, **kwargs):
-        elif self._state_dict_type == StateDictType.LOCAL_STATE_DICT:
+        elif (
+            self._state_dict_type == StateDictType.LOCAL_STATE_DICT or
+            self._state_dict_type == StateDictType.SHARDED_STATE_DICT
```
so it seems that sharded state dict calling `state_dict` is not meant to do much, we remove the checkpointed FLAT_PARAM and add new data to the state_dict for the sharded original parameters. 

If this is the case, could we just remove the super().state_dict calls, recurse ourselves and call the post hook?

fegin replied:
We need the recursive calls for 1.) constructing the correct prefix, 2.) calling post_hook in a reversed order. We can do this ourselves but it will be better to reuse state_dict logic. The only thing that sharded_state_dict does not need is the detach one but that should not causing too many overheads.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1983,19 +1983,25 @@ def _sharded_post_state_dict_hook(
+            "not be SUMMON_FULL_PARAMS."
+        )
+        with self._summon_full_params(recurse=False, writeback=False):
+            for fqn, _, _ in self._param_fqns:
```
wouldn't `named_parameters()` also just give the parameter names we need?

fegin replied:
named_parameters() will give us more parameters than what we need. We need to use recursive named_parameters() and this will give us more parameters: 1.) parameters in children FSDP modules 2.) parameters that are ignored by FSDP.

rohan-varma reviewed (approved):
Thanks for the fix! It would be great to test it on a use case where full state dict fails but this one succeeds.

fegin committed:
Update on "[FSDP] Move the sharded_state_dict logic to the post hook to avoid OOM"


The original implementation put the call of `_summon_full_params()` in `state_dict()`. However, because `state_dict()` is recursive, `_summon_full_params()` will also behave like the recursive version even if recursive is set to False. This PR put the logic in the post hook to solve the OOM issue.

Differential Revision: [D38329396](https://our.internmc.facebook.com/intern/diff/D38329396/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
